### PR TITLE
Relationals (Eq, Ne, Ge, Gt, Le, Lt)

### PR DESCRIPTION
**Pull request by ShikharJ**
TODO:
- [x] Write test-cases

Relevant: #1087

### CODE REVIEW

File: symengine/printer.cpp [hunk 1]
```diff
@@ -136,6 +136,48 @@ void StrPrinter::bvisit(const ComplexDouble &x)
...
+}
+
+void StrPrinter::bvisit(const GreaterThan &x)
+{
+    std::ostringstream s;
+    s << apply(x.get_lhs()) << " >= " << apply(x.get_rhs());
+    str_ = s.str();
+}
+
+void StrPrinter::bvisit(const LessThan &x)
```
Comment (isuruf) [inline symengine/printer.cpp]: I don't think we need both `GreaterThan` and `LessThan`. Having one is easier.

File: symengine/relationals.cpp [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+    SYMENGINE_ASSERT(is_a<Relational>(o));
+    const Relational &r = down_cast<const Relational &>(o);
+    int cmp = lhs_->__cmp__(*(r.get_lhs()));
+    if (cmp != 0)
+        return cmp;
+    cmp = rhs_->__cmp__(*(r.get_rhs()));
+    return cmp;
+}
+
+RCP<const Basic> Relational::canonical() const
```
Comment (isuruf) [inline symengine/relationals.cpp]: In SymEngine, we keep all of the objects in the canonical form. I would change the design to create a `Relational` only in the canonical form

File: symengine/relationals.cpp [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+}
+
+inline RCP<const Basic> GreaterThan::reversed() const
+{
+    return make_rcp<const LessThan>(get_rhs(), get_lhs());
+}
+
+inline bool GreaterThan::eval_relation() const
+{
+    int cmp = get_lhs()->__cmp__(*get_rhs());
```
Comment (isuruf) [inline symengine/relationals.cpp]: `__cmp__` is structural comparison. Don't use it for mathematical comparison

File: symengine/relationals.cpp (hunk 2 above)
Comment (ShikharJ) [inline symengine/relationals.cpp]: Canonicalization is achieved through a priority basis here. I'm doubtful how that would be incorporated into the current design efficiently. Moreover, this function is for `SymPy` compatibility.

File: symengine/relationals.cpp (hunk 3 above)
Comment (ShikharJ) [inline symengine/relationals.cpp]: What can be used for mathematical comparisons?

**Comment by ShikharJ** [comment]
@isuruf Would you prefer the proposed changes? Or should I revert?

**Comment by isuruf** [comment]
It's better now. Still, there are both `GreaterThan` and `LesserThan` classes. You should remove one of those classes.

**Comment by ShikharJ** [comment]
This PR is ready for a review @isuruf @Sumith1896.

**Comment by isuruf** [comment]
Can you remove `StrictGreaterThan` as well?

**Comment by isuruf** [comment]
Also, You were right about `TwoArgFunction`. We can avoid code duplication if we inherit from `TwoArgFunction` as well.

**Comment by ShikharJ** [comment]
Ping @isuruf.

File: symengine/functions.cpp [hunk 4]
```diff
@@ -1065,7 +1065,7 @@ RCP<const Basic> acot(const RCP<const Basic> &arg)
 }
 
 ATan2::ATan2(const RCP<const Basic> &num, const RCP<const Basic> &den)
-    : TwoArgFunction(num, den)
+    : TwoArgBasic<Function>(num, den)
```
Comment (isuruf) [inline symengine/functions.cpp]: Use `typedef TwoArgBasic<Function> TwoArgFunction` in the header file. Then there's no need to change these

File: symengine/logic.cpp [hunk 5]
```diff
@@ -478,6 +478,202 @@ RCP<const Boolean> logical_xor(const vec_boolean &s)
     }
 }
 
+Relational::Relational(const RCP<const Basic> &lhs, const RCP<const Basic> &rhs)
+    : TwoArgBasic<Boolean>(lhs, rhs)
+{
+}
+
+hash_t Relational::__hash__() const
```
Comment (isuruf) [inline symengine/logic.cpp]: Remove this. This is defined by `TwoArgBasic`

File: symengine/logic.cpp [hunk 6]
```diff
@@ -478,6 +478,202 @@ RCP<const Boolean> logical_xor(const vec_boolean &s)
...
+{
+    hash_t seed = this->get_type_code();
+    hash_combine<Basic>(seed, *get_arg1());
+    hash_combine<Basic>(seed, *get_arg2());
+    return seed;
+}
+
+inline bool Relational::__eq__(const Basic &o) const
```
Comment (isuruf) [inline symengine/logic.cpp]: Remove this. This is declared in `TwoArgBasic`

File: symengine/logic.cpp [hunk 7]
```diff
@@ -478,6 +478,202 @@ RCP<const Boolean> logical_xor(const vec_boolean &s)
...
+                        or (eq(*get_arg1(), *(s.get_arg2()))
+                            and eq(*get_arg2(), *(s.get_arg1()))));
+            return (eq(*get_arg1(), *(s.get_arg1()))
+                    and eq(*get_arg2(), *(s.get_arg2())));
+        }
+    }
+    return false;
+}
+
+inline int Relational::compare(const Basic &o) const
```
Comment (isuruf) [inline symengine/logic.cpp]: Remove this as well

File: symengine/logic.h [hunk 8]
```diff
@@ -168,6 +168,97 @@ class Xor : public Boolean
...
+//! Returns the canonicalized Equality object from the two arguments
+RCP<const Basic> Eq(const RCP<const Basic> &lhs);
+//! Returns the canonicalized Equality object from the two arguments
+RCP<const Basic> Eq(const RCP<const Basic> &lhs, const RCP<const Basic> &rhs);
+//! Returns the canonicalized Unequality object from the arguments
+RCP<const Basic> Ne(const RCP<const Basic> &lhs, const RCP<const Basic> &rhs);
+//! Returns the canonicalized LessThan object from the arguments
+RCP<const Basic> Le(const RCP<const Basic> &lhs, const RCP<const Basic> &rhs);
+//! Returns the canonicalized StrictLessThan object from the arguments
+RCP<const Basic> Lt(const RCP<const Basic> &lhs, const RCP<const Basic> &rhs);
```
Comment (isuruf) [inline symengine/logic.h]: Can you create `Gt` and `Ge` convenience functions?

File: symengine/logic.h [hunk 9]
```diff
@@ -168,6 +168,97 @@ class Xor : public Boolean
...
+
+inline bool is_a_Relational(const Basic &b)
+{
+    return (b.get_type_code() == EQUALITY || b.get_type_code() == UNEQUALITY
+            || b.get_type_code() == LESSTHAN
+            || b.get_type_code() == STRICTLESSTHAN);
+}
+
+//! Returns the canonicalized Equality object from the two arguments
+RCP<const Basic> Eq(const RCP<const Basic> &lhs);
```
Comment (isuruf) [inline symengine/logic.h]: What does this do?

File: symengine/printer.cpp [hunk 10]
```diff
@@ -136,6 +136,34 @@ void StrPrinter::bvisit(const ComplexDouble &x)
...
+{
+    std::ostringstream s;
+    s << "Le(" << apply(x.get_arg1()) << ", " << apply(x.get_arg2()) << ")";
+    str_ = s.str();
+}
+
+void StrPrinter::bvisit(const StrictLessThan &x)
+{
+    std::ostringstream s;
+    s << "Lt(" << apply(x.get_arg1()) << ", " << apply(x.get_arg2()) << ")";
```
Comment (isuruf) [inline symengine/printer.cpp]: I think `x < y` syntax is easier to read.

File: symengine/logic.h (hunk 9 above)
Comment (ShikharJ) [inline symengine/logic.h]: Its a convenience function, when only a single argument is provided, automatically taking the `rhs` to be `zero`.

File: symengine/logic.cpp (hunk 6 above)
Comment (ShikharJ) [inline symengine/logic.cpp]: This has to be overloaded, as the case `eq(Eq(x, y), Eq(y, x))` is not considered in `TwoArgBasic`.

Comment (isuruf) [inline symengine/logic.cpp]: Then you have to change your implementation of hash.

Comment (isuruf) [inline symengine/logic.cpp]: Better solution is to do what we did in the `beta` function. `Eq(y, x)` would return `Equality(x, y)`

**Comment by ShikharJ** [comment]
Ping @isuruf.

File: symengine/logic.cpp [hunk 11]
```diff
@@ -478,6 +478,192 @@ RCP<const Boolean> logical_xor(const vec_boolean &s)
...
+
+RCP<const Basic> Gt(const RCP<const Basic> &lhs, const RCP<const Basic> &rhs)
+{
+    RCP<const Basic> r = Le(lhs, rhs);
+    if (is_a<BooleanAtom>(*r)) {
+        if (eq(*boolTrue, *r))
+            return boolFalse;
+        return boolTrue;
+    }
+    return make_rcp<StrictLessThan>(mul(lhs, minus_one), mul(rhs, minus_one));
```
Comment (isuruf) [inline symengine/logic.cpp]: You can just do `return Lt(rhs, lhs)`

File: symengine/logic.cpp [hunk 12]
```diff
@@ -478,6 +478,192 @@ RCP<const Boolean> logical_xor(const vec_boolean &s)
...
+    return Ne(lhs, rhs);
+}
+
+RCP<const Basic> Ne(const RCP<const Basic> &lhs, const RCP<const Basic> &rhs)
+{
+    RCP<const Basic> r = Eq(lhs, rhs);
+    if (is_a<BooleanAtom>(*r)) {
+        if (eq(*boolTrue, *r))
+            return boolFalse;
+        return boolTrue;
```
Comment (isuruf) [inline symengine/logic.cpp]: Just do `return r;` instead of the above 3 lines

File: symengine/logic.cpp [hunk 13]
```diff
@@ -478,6 +478,192 @@ RCP<const Boolean> logical_xor(const vec_boolean &s)
...
+}
+
+RCP<const Basic> Eq(const RCP<const Basic> &lhs, const RCP<const Basic> &rhs)
+{
+    if (is_a<NaN>(*lhs) or is_a<NaN>(*rhs))
+        return boolean(false);
+    bool b = eq(*lhs, *rhs);
+    if (b)
+        return boolean(true);
+    else {
```
Comment (isuruf) [inline symengine/logic.cpp]: See https://github.com/symengine/symengine/blob/master/doc/style_guide.md#whitespace

File: symengine/logic.cpp (hunk 12 above)
Comment (ShikharJ) [inline symengine/logic.cpp]: But shouldn't the complement of the `BooleanAtom` be returned here?

Comment (isuruf) [inline symengine/logic.cpp]: Right. then `return logical_not(r);`

**Comment by ShikharJ** [comment]
Ping @isuruf.

File: symengine/logic.cpp [hunk 14]
```diff
@@ -478,6 +478,171 @@ RCP<const Boolean> logical_xor(const vec_boolean &s)
...
+    }
+    return make_rcp<const StrictLessThan>(lhs, rhs);
+}
+
+RCP<const Basic> Gt(const RCP<const Basic> &lhs, const RCP<const Basic> &rhs)
+{
+    RCP<const Basic> r = Le(lhs, rhs);
+    if (is_a<BooleanAtom>(*r)) {
+        return logical_not(rcp_static_cast<const BooleanAtom>(r));
+    }
```
Comment (isuruf) [inline symengine/logic.cpp]: Above 4 lines are unnecessary

File: symengine/logic.cpp [hunk 15]
```diff
@@ -478,6 +478,171 @@ RCP<const Boolean> logical_xor(const vec_boolean &s)
...
+    return make_rcp<const LessThan>(lhs, rhs);
+}
+
+RCP<const Basic> Ge(const RCP<const Basic> &lhs, const RCP<const Basic> &rhs)
+{
+    RCP<const Basic> r = Lt(lhs, rhs);
+    if (is_a<BooleanAtom>(*r)) {
+        return logical_not(rcp_static_cast<const BooleanAtom>(r));
+    }
+    return make_rcp<LessThan>(mul(lhs, minus_one), mul(rhs, minus_one));
```
Comment (isuruf) [inline symengine/logic.cpp]: Instead of above 5 lines, use `return Le(rhs, lhs)`

**Comment by ShikharJ** [comment]
Ping @isuruf.

**Comment by isuruf** [comment]
Looks good to me. We also need to add these to `SubsVisitor`, `logical_not` and `CCodePrinter`.

**Pull request merged**